Objects/typeobject.c

```diff
@@ -2342,9 +2349,8 @@ int
 PyType_IsSubtype(PyTypeObject *a, PyTypeObject *b)
 {
 #ifdef Py_GIL_DISABLED
-    PyObject *mro = _PyType_GetMRO(a);
+    PyObject *mro = _Py_atomic_load_ptr(&a->tp_mro);
```

Comment:
Let's simplify `PyType_IsSubtype` and also `_PyType_GetMRO` to only use what's now the code path for the default build. In other words, for `PyType_IsSubtype` just:

```
   return is_subtype_with_mro(lookup_tp_mro(a), a, b);
```

We'll also need to slightly modify `lookup_tp_mro` to remove the type lock assertion and use `FT_ATOMIC_LOAD_PTR`.

Reply:
I'm a little unclear about what's safe to do with QSBR tagged memory. I know that reads are fine, but are writes fine too? In this particular case, is it safe to replace 

```
    if (mro == NULL) {
        return NULL;
    }
    if (_Py_TryIncref(&self->tp_mro, mro)) {
        return mro;
    }
    BEGIN_TYPE_LOCK();
    mro = lookup_tp_mro(self);
    Py_XINCREF(mro);
    END_TYPE_LOCK()
    return mro;
```

in `PyType_GetMRO` with just `Py_XINCREF(mro)`?

Reply:
Sorry - I was confused. I thought we were delaying the `Py_DECREF`, but we are only delaying the `PyObject_Free()`.

To answer your question:
1) `_PyType_GetMRO()` is never used and not exported, so the implementation doesn't particularly matter. Let's stick with the simple implementation of `Py_XINCREF(mro)`. We can consider deleting the function in a later PR, but since it predates the free-threading changes, I wouldn't want to delete it here (just simplify it).
2) `Py_XINCREF(obj) would generally not be safe because the object may have zero refcount and it could resurrect a dead object in an unsafe way.

I think we really want to delay the decref or at least the possible destructor call, not just the `PyObject_Free`. We don't need to use QSBR, we could delay it until the next GC if that's simpler to implement.

Reply:
> I think we really want to delay the decref or at least the possible destructor call, not just the PyObject_Free.

Using QSBR to delay freeing the tuple should work for this particular use case, since `PyType_IsSubtype` only needs to perform pointer equality checks on the contents of the tuple, and the tuple isn't cleared in the destructor. That said, we could run into issues if the implementation of `tupledealloc` changed.

> I think we really want to delay the decref or at least the possible destructor call, not just the PyObject_Free. We don't need to use QSBR, we could delay it until the next GC if that's simpler to implement.

This would definitely be more robust. Is there an existing mechanism (like how QSBR lets us delay freeing the memory) for doing this?

Reply:
> Is there an existing mechanism (like how QSBR lets us delay freeing the memory) for doing this?

No. I'm okay with relying on the particular behavior for now as long as we track it in an issue. We can address it later on before the 3.13 release.